src/rust.rs

```diff
@@ -1257,3 +1258,73 @@ macro_rules! new_jsjitinfo_bitfield_1 {
             (($slotIndex as u32) << 22u32)
     }
 }
+
+pub struct CapturedJSStack<'a> {
+    cx: *mut JSContext,
+    stack: RootedGuard<'a, *mut JSObject>,
+}
+
+impl<'a> CapturedJSStack<'a> {
+    pub fn new(cx: *mut JSContext, mut guard: RootedGuard<'a, *mut JSObject>, max_frame_count: u32) -> Option<Self> { // or maybe another name, for example "capture"?
+        let obj_handle = guard.handle_mut(); // if nullptr, return None
+                                             // how to check if handle is not nullptr?
+                                             // how to handle such errors in a nice way?
+        unsafe {
+            if !CaptureCurrentStack(cx, obj_handle, max_frame_count) {
+                return None;
+            }
+        }
+
+        Some(CapturedJSStack {
+            cx: cx,
+            stack: guard,
+        })
+    }
+
+    pub fn as_string(&self, indent: usize) -> Option<String> {
+        // TODO where errors can occur? some example of non trivial error handling?
+        // TODO how to write tests for this code? Including the problem of obtaining some stack trace
+        // do we want to do the actual logic of conversion in this method or maybe in constructor and just return cached result?
+        unsafe { // since almost everything here is unsafe, is it okay to wrap it like this?
+            let stack_handle = self.stack.handle();
+            rooted!(in(self.cx) let mut js_string = JS_NewStringCopyZ(self.cx, ptr::null()));
+            let string_handle = js_string.handle_mut(); // as above -- how to check if is not nullptr?
+
+            if !BuildStackString(self.cx, stack_handle, string_handle, indent) {
+                return None;
+            }
+
+            // wanted somehow to just get the utf-16 string and use String::from_utf16_lossy
+            // but JS_GetTwoByteStringCharsAndLength requires some object of strange type (AutoCheckCannotGC)
```

Comment:
https://github.com/servo/rust-mozjs/blob/2b49303d32598a4b00010d6d159ec728ef0145e3/src/conversions.rs#L515-L519

Reply:
Are you sure about passing null pointer to the `JS_GetTwoByteStringCharsAndLength`? In C++ it requires a reference: https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/JSAPI_Reference/JS_GetLatin1StringCharsAndLength. Maybe it doesn't have any internal fields, but isn't it undefined behavior?

I read the comment before definition (https://dxr.mozilla.org/mozilla-central/source/js/public/GCAPI.h?q=AutoCheckCannotGC&redirect_type=direct#852-876), but I am still not sure when we can use it. Does it mean that we can use it when we have rooted all the objects whose handles we are passing to the function?

Reply:
If you look at where the `nogc` parameter is actually used in the function body, you'll see that it's passed as a parameter to [twoByteChars](https://dxr.mozilla.org/mozilla-central/source/js/src/vm/String.h#681), and that function doesn't do anything with it, so I'm quite confident that there's no UB going on here.

In fact, the comment before the definition of AutoCheckCannotGC states that this is only used for static rooting hazard analysis, so I don't see how it can go wrong when you send in a null pointer.

Reply:
@KiChjang thank you for the explanation. I wasn't sure.